### PR TITLE
Initial commit for 2.12 support

### PR DESCRIPTION
**Pull request by mdedetrich**

### CODE REVIEW

**Comment by omarkilani** [comment]
Hi @mdedetrich,

There's a 2.12 effort in progress at https://github.com/hamnis/unfiltered/commit/fbd78b9fde45976a5ba759f181e191a01a5366b6

**Comment by mdedetrich** [comment]
I have just merged with upstream, still being blocked by https://github.com/softprops/tubesocks/issues/6 ?

From what I see, the main difference is that I have updated the tests for ScalaTest 3.x

**Comment by omarkilani** [comment]
There seems to be some sort of non-deterministic error when running 'sbt test'.

Could you please add the following to your branch while you're there? :)

```patch
diff --git a/.travis.yml b/.travis.yml
index 60fb030..e6b592e 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -3,6 +3,7 @@ sudo: false
 scala:
   - 2.10.6
   - 2.11.8
+  - 2.12.0
 jdk:
   - oraclejdk7
   - oraclejdk8
diff --git a/library/build.sbt b/library/build.sbt
index 8964e56..cee9e50 100644
--- a/library/build.sbt
+++ b/library/build.sbt
@@ -7,9 +7,9 @@ unmanagedClasspath in (LocalProject("unfiltered"), Test) ++= {
 }
 
 libraryDependencies ++= Seq(
-  "commons-codec" % "commons-codec" % "1.4",
+  "commons-codec" % "commons-codec" % "1.10",
   Common.specs2Dep(scalaVersion.value) % "test",
-  "org.scalacheck" %% "scalacheck" % "1.12.6" % "test",
+  "org.scalacheck" %% "scalacheck" % "1.13.4" % "test",
   "joda-time" % "joda-time" % "2.9.6" % "test",
   "org.joda" % "joda-convert" % "1.8.1" % "test"
 )
```

**Comment by mdedetrich** [comment]
Done

**Comment by omarkilani** [comment]
Alright great. That's looking pretty good. Guess we're blocked on tubesocks.

I'm not sure how to modify the travis.yml to only run the 2.12 tests with jdk8 without some hackery.

**Comment by omarkilani** [comment]
Looking into travis stuff, this *should* work:

```patch
diff --git a/.travis.yml b/.travis.yml
index e6b592e..ddfa252 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -7,3 +7,8 @@ scala:
 jdk:
   - oraclejdk7
   - oraclejdk8
+matrix:
+  exclude:
+    -
+      scala: 2.12.0
+      jdk: oraclejdk7
```

There appears to be an issue in travis-yaml which generates some warnings with this YAML.

There are multiple instances of other people running into this issue:

https://github.com/travis-ci/travis.rb/issues/376
https://github.com/travis-ci/travis-ci/issues/6540

Not entirely clear if this will just work or not.

**Comment by mdedetrich** [comment]
Okay, pushed the patch again

**Comment by omarkilani** [comment]
That looks like some sort of race condition in the tests or some OS issue where the listening port isn't freed fast enough to run the next test. I think the switch to okhttp3 from dispatch has made the tests run faster than they used to, or something like that.

At least the travis thing seems to have worked. :)

**Comment by omarkilani** [comment]
Re tubesocks, since we're now using okhttp3 to run other tests, we could switch the websocket tests to use that instead.

**Comment by omarkilani** [comment]
I think you need to bring over the 'BeforeAndAfterAll' parts of https://github.com/hamnis/unfiltered/commit/fbd78b9fde45976a5ba759f181e191a01a5366b6

For these tests to work properly.

File: project/common.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ object Common {
 
     version := "0.9.0-beta1",
 
-    crossScalaVersions := Seq("2.11.8", "2.10.4"),
+    crossScalaVersions := Seq("2.11.8", "2.10.6", "2.12.0"),
```
Comment (omarkilani) [inline project/common.scala]: Could you please reorder this so that 2.12.0 is first in the list?

**Comment by xuwei-k** [comment]
> Re tubesocks, since we're now using okhttp3 to run other tests, we could switch the websocket tests to use that instead.


There is another way

```diff
diff --git a/netty-websockets/build.sbt b/netty-websockets/build.sbt
index bddd1d5..5f29b77 100644
--- a/netty-websockets/build.sbt
+++ b/netty-websockets/build.sbt
@@ -4,5 +4,3 @@ unmanagedClasspath in (local("netty-websockets"), Test) ++=
   (fullClasspath in (local("specs2"), Compile)).value
 
 libraryDependencies ++= Common.integrationTestDeps(scalaVersion.value)
-
-libraryDependencies += "me.lessis" %% "tubesocks" % "0.1.0" % "test"// exclude("io.netty", "netty")
diff --git a/project/build.scala b/project/build.scala
index d5bd713..bff3b7c 100644
--- a/project/build.scala
+++ b/project/build.scala
@@ -97,7 +97,9 @@ object Unfiltered extends Build {
     ).dependsOn(library)
 
   lazy val websockets =
-    module("netty-websockets")().dependsOn(nettyServer)
+    module("netty-websockets")().dependsOn(nettyServer).dependsOn(
+      RootProject(uri("git://github.com/xuwei-k/tubesocks.git#7c0b961e7d077381e2a7581d9d8b6a975a3a417f")) % "test"
+    )
```

https://github.com/xuwei-k/tubesocks/commit/7c0b961e7d077381e2a7581d9d8b6a975a3a417f

**Comment by omarkilani** [comment]
Or we can just pull the 3 scala files from tubesocks into one tubesocks.scala file and put it next to the test? :)

One less dependency to worry about...

File: project/common.scala (hunk 1 above)
Comment (mdedetrich) [inline project/common.scala]: Done

**Comment by mdedetrich** [comment]
@omarkilani Maybe lets wait a couple of days for tubesocks. I will see if I can contact the maintainer directly

**Pull request merged**